src/main/java/com/rackspace/jenkins_nodepool/HoldStep.java

```diff
@@ -0,0 +1,147 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ *
+ * @author Rackspace
+ */
+public class HoldStep extends Builder implements SimpleBuildStep {
+
+    private static final Logger LOG = Logger.getLogger(HoldStep.class.getName());
+
+    private final String reason;
+    private final String duration;
+    private PrintStream consoleLog;
+
+    @DataBoundConstructor
+    public HoldStep(String duration, String reason){
+        this.reason = reason;
+        this.duration = duration;
+    }
+
+    public HoldStep(String duration){
+        this(duration, "Held from Pipeline Step");
+    }
+
+    public HoldStep(){
+        this("1d");
+    }
+
+    public String getReason(){
+        return reason;
+    }
+    public String getDuration(){
+        return duration;
+    }
+
+    private void log(String message, Level level){
+        consoleLog.println(message);
+        LOG.log(level, message);
+    }
+    private void log(String message){
+        log(message, Level.INFO);
+    }
+    private void setConsoleLogger(PrintStream logger){
+        this.consoleLog = logger;
+    }
+
+    @Override
+    public void perform(Run<?, ?> run, FilePath workspace, Launcher launcher, TaskListener listener) throws InterruptedException, IOException {
+        setConsoleLogger(listener.getLogger());
+        // try and find the current computer
+        // use a few different methods so freestyle and pipeline jobs work
+        Executor executor = Executor.currentExecutor();
+        VirtualChannel vc = launcher.getChannel();
+        FilePath fp = new FilePath(vc, "/");
+        Job job = run.getParent();
+
+        Computer computer;
+        if (executor != null){
+            computer = executor.getOwner();
+        } else if (Computer.currentComputer() != null) {
+            computer = Computer.currentComputer();
+        } else if(fp.toComputer() != null){
+            computer = fp.toComputer();
+        } else {
+            log("Can't hold as theres no node allocated :(");
+            return;
+        }
+
+        if (computer instanceof NodePoolComputer){
+            try {
+                NodePoolComputer npc = (NodePoolComputer) computer;
+                NodePoolSlave nps = (NodePoolSlave) npc.getNode();
+                if (nps != null){
+                    String build_id = job.getDisplayName()+"-"+run.getNumber();
+                        nps.setHeld(true);
+                        nps.setHoldReason(reason);
+                        nps.setHoldUser(build_id);
```

Comment:
Is this the best value for hold user? Based on the above if/else logic to get the computer, do we have access to the owner? If nothing is appropriate, then build ID is probably our best bet.

Reply:
I think so, it could possible be the pull request author, but then I thought it would be confusing as to whether that user manually created the hold or whether it happened as part of the job.